plugin/session_buffer.py

```diff
@@ -484,21 +489,21 @@ def do_semantic_tokens_async(self, view: sublime.View, only_viewport: bool = Fal
         params = {"textDocument": text_document_identifier(view)}  # type: Dict[str, Any]
         if only_viewport and self.session.has_capability("semanticTokensProvider.range"):
             params["range"] = region_to_range(view, view.visible_region())
-            request = Request.semanticTokensRange(params, view)
+            request = Request.semanticTokensRange(cast(SemanticTokensRangeParams, params), view)
```

Comment:
We could avoid the heavy-handed `cast` if we allow our selfs a little more code duplication.

```py
            params = {
                "textDocument": text_document_identifier(view),
                "range": region_to_range(view, view.visible_region()),
            }  # type: SemanticTokensRangeParams
            request = Request.semanticTokensRange(params, view)
```

Reply:
I can't get this to work (unless I would use different variable names instead of `params` for each `if` branch - which is awkward).

It's like if Pyright doesn't understand that `if ... elif` branches are mutually exclusive 😕 

![pyright](https://user-images.githubusercontent.com/6579999/210697402-91ccfbe4-33b1-4d06-b9c8-d6f0cedabd10.png)

<details><summary>Code:</summary>
<p>

```python
        if only_viewport and self.session.has_capability("semanticTokensProvider.range"):
            params = {
                "textDocument": text_document_identifier(view),
                "range": region_to_range(view, view.visible_region())
            }  # type: SemanticTokensRangeParams
            request = Request.semanticTokensRange(params, view)
            self.semantic_tokens.pending_response = self.session.send_request_async(
                request, partial(self._on_semantic_tokens_viewport_async, view), self._on_semantic_tokens_error_async)
        elif self.semantic_tokens.result_id and self.session.has_capability("semanticTokensProvider.full.delta"):
            params = {
                "textDocument": text_document_identifier(view),
                "previousResultId": self.semantic_tokens.result_id
            }  # type: SemanticTokensDeltaParams
            request = Request.semanticTokensFullDelta(cast(SemanticTokensDeltaParams, params), view)
            self.semantic_tokens.pending_response = self.session.send_request_async(
                request, self._on_semantic_tokens_delta_async, self._on_semantic_tokens_error_async)
        elif self.session.has_capability("semanticTokensProvider.full"):
            params = cast(SemanticTokensParams, {"textDocument": text_document_identifier(view)})
            request = Request.semanticTokensFull(params, view)
            self.semantic_tokens.pending_response = self.session.send_request_async(
                request, self._on_semantic_tokens_async, self._on_semantic_tokens_error_async)
        elif self.session.has_capability("semanticTokensProvider.range"):
            params = {
                "textDocument": text_document_identifier(view),
                "range": region_to_range(view, view.visible_region())
            }  # type: SemanticTokensRangeParams
            params["range"] = entire_content_range(view)
            request = Request.semanticTokensRange(params, view)
            self.semantic_tokens.pending_response = self.session.send_request_async(
                request, self._on_semantic_tokens_async, self._on_semantic_tokens_error_async)
```
</p>
</details>

Reply:
Hmm, OK, that's unfortunate.

Reply:
Could inline the params like this though:

```py
        self.semantic_tokens.view_change_count = view.change_count()
        if only_viewport and self.session.has_capability("semanticTokensProvider.range"):
            request = Request.semanticTokensRange(
                {
                    "textDocument": text_document_identifier(view),
                    "range": region_to_range(view, view.visible_region())
                }, view)
            self.semantic_tokens.pending_response = self.session.send_request_async(
                request, partial(self._on_semantic_tokens_viewport_async, view), self._on_semantic_tokens_error_async)
        elif self.semantic_tokens.result_id and self.session.has_capability("semanticTokensProvider.full.delta"):
            request = Request.semanticTokensFullDelta(
                {
                    "textDocument": text_document_identifier(view),
                    "previousResultId": self.semantic_tokens.result_id
                }, view)
            self.semantic_tokens.pending_response = self.session.send_request_async(
                request, self._on_semantic_tokens_delta_async, self._on_semantic_tokens_error_async)
        elif self.session.has_capability("semanticTokensProvider.full"):
            request = Request.semanticTokensFull(
                {
                    "textDocument": text_document_identifier(view)
                }, view)
            self.semantic_tokens.pending_response = self.session.send_request_async(
                request, self._on_semantic_tokens_async, self._on_semantic_tokens_error_async)
        elif self.session.has_capability("semanticTokensProvider.range"):
            request = Request.semanticTokensRange(
                {
                    "textDocument": text_document_identifier(view),
                    "range": entire_content_range(view)
                }, view)
            self.semantic_tokens.pending_response = self.session.send_request_async(
                request, self._on_semantic_tokens_async, self._on_semantic_tokens_error_async)
```